lib/common/common/src/tar_ext.rs

```diff
@@ -13,16 +14,92 @@ use tokio::task::JoinError;
 /// 2. Provides the [`BuilderExt::descend`] method.
 #[derive(Clone)]
 pub struct BuilderExt {
-    tar: Arc<Mutex<tar::Builder<File>>>,
+    tar: Arc<Mutex<BuilderBox>>,
     path: PathBuf,
 }
 
+/// A wrapper around [`tar::Builder<WriteBox>`] that disables [`WriteBox`] when
+/// it is dropped.
+///
+/// Disabling the [`WriteBox`] is a workaround for the inconvenient
+/// [`tar::Builder`] behavior: dropping a [`tar::Builder`] might cause a final
+/// write of archive footer.
+/// This behavior is problematic for [`Write`] implementations that could panic
+/// when used in an async context, such as [`SyncIoBridge`].
+///
+/// [`SyncIoBridge`]: https://docs.rs/tokio-util/0.7.12/tokio_util/io/struct.SyncIoBridge.html#method.new
+struct BuilderBox {
+    tar: Option<tar::Builder<WriteBox>>,
+    enabled: Arc<AtomicBool>,
+}
+
+/// A wrapper around [`Write`] that could be disabled by [`BuilderBox`].
+struct WriteBox {
+    inner: Box<dyn Send + WriteSeek>,
+    enabled: Arc<AtomicBool>,
+}
+
+/// TODO: This trait will be replaced with just [`Write`] once we get rid of
+/// nested tar archives.
+trait WriteSeek: Write + Seek {}
+impl<T: Write + Seek> WriteSeek for T {}
+
+impl BuilderBox {
+    fn tar(&mut self) -> &mut tar::Builder<WriteBox> {
+        self.tar.as_mut().unwrap()
+    }
+}
+
+impl Drop for BuilderBox {
+    fn drop(&mut self) {
+        self.enabled.store(false, Ordering::Relaxed);
+    }
+}
+
+impl Write for WriteBox {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        if !self.enabled.load(Ordering::Relaxed) {
```

Comment:
Do I understand it correctly: the tar impl will attempt to write the final 1K zero bytes. But this write will be rejected here if we unset the enabled flag, preventing the panic?

If we explicitly finalize ourselves the tar impl doesn't write anything and we don't hit this branch at all (during drop).

Reply:
Yes, that's correct.